### PR TITLE
Migration from TSLint to ESLint

### PR DESCRIPTION
**Pull request by darzu**
# Migration from TSLint to ESLint

I primarily followed: https://code.visualstudio.com/api/advanced-topics/tslint-eslint-migration.

I used "tslint-to-eslint-config" CLI tool to generate a .eslintrc file using our tslint.json, however there were lots of rules that tslint-to-eslint-config didn't know about.

Some of the gaps from "tslint-microsoft-contrib" where filled in by "@microsoft/eslint-plugin-sdl" which is sort of its successor.

Some of the gaps from "react-a11y" tslint plugin where filled by "jsx-a11y".

Bizarrely, a couple of files (blocks.tsx and snippetBuilder.tsx; see .eslintignore) cause the ESLint to crash deep inside the TS compiler ☹️. So I had to disable these files from linting for now, but I suspect we just need to update our TS version (future PR).

## Steps for Devs
After this PR is merged, do the following in pxt (no other repo should be affected):
0. `git pull`
1. `rm -rf node_modules/`
2. `npm install`
3. in vscode, uninstall or disable tslint extension (if you want)
4. in vscode, install "dbaeumer.vscode-eslint" extension
5. run `gulp && gulp lint` to make sure everything works
6. in vscode, inside a file (say webapp/src/app.tsx), try adding some trailing whitespace. You should see this squigly error: `Trailing spaces not allowed. eslintno-trailing-spaces`
7. you're done!

## Removed rules
These rules where removed and tslint-to-eslint-config couldn't find replacements:
<details>
    <summary>Expand</summary>

 - "jquery-deferred-must-complete"
 - "mocha-avoid-only"
 - "mocha-no-side-effect-code"
 - "no-delete-expression"
 - "no-disable-auto-sanitization"
 - "no-document-domain"
 - "no-document-write"
 - "no-exec-script"
 - "no-function-constructor-with-string-args"
 - "no-http-string"
 - "no-jquery-raw-elements"
 - "no-string-based-set-immediate"
 - "no-string-based-set-interval"
 - "no-string-based-set-timeout"
 - "no-unnecessary-bind"
 - "no-unnecessary-override"
 - "no-unsupported-browser-code"
 - "no-with-statement"
 - "non-literal-require"
 - "possible-timing-attack"
 - "react-anchor-blank-noopener"
 - "react-no-dangerous-html"
 - "react-this-binding-issue"
 - "react-unused-props-and-state"
</details>
<br>

These rules where technically replaced but behaved differently enough that I had to disable them:
 - "no-underscore-dangle"
 - "brace-style"

These rules could be supported if we used the eslint-tslint crossover plugin but the whole point of this PR is expunge tslint:
 -  "match-default-export-name": true,
 -  "whitespace"

## tslint-microsoft-contrib -> eslint-plugin-sdl
As mentioned, many of the tslint-microsoft-contrib rules where replaced by ones from the new MSFT SDL plugin. However there's still many more rules that we need to enable for full SDL compliance. That's out of scope of this PR.

See:
 - https://github.com/microsoft/eslint-plugin-sdl

## react-a11y -> jsx-a11y
Our "react-a11y" rules where sort of replaced with these other rules from the "jsx-a11y" plugin:

Removed react-a11y rules:
<details>
    <summary>Expand</summary>

 - "react-a11y-anchors"
 - "react-a11y-aria-unsupported-elements"
 - "react-a11y-event-has-role"
 - "react-a11y-image-button-has-alt"
 - "react-a11y-img-has-alt"
 - "react-a11y-lang"
 - "react-a11y-meta"
 - "react-a11y-props"
 - "react-a11y-proptypes"
 - "react-a11y-role"
 - "react-a11y-role-has-required-aria-props"
 - "react-a11y-role-supports-aria-props"
 - "react-a11y-tabindex-no-positive"
 - "react-a11y-titles"
</details>
<br>

New jsx-a11y rules:
<details>
    <summary>Expand</summary>

 - "jsx-a11y/accessible-emoji"
 - "jsx-a11y/alt-text"
 - "jsx-a11y/aria-activedescendant-has-tabindex"
 - "jsx-a11y/aria-props"
 - "jsx-a11y/aria-role"
 - "jsx-a11y/aria-unsupported-elements"
 - "jsx-a11y/autocomplete-valid"
 - "jsx-a11y/heading-has-content"
 - "jsx-a11y/html-has-lang"
 - "jsx-a11y/img-redundant-alt"
 - "jsx-a11y/no-access-key"
 - "jsx-a11y/no-distracting-elements"
 - "jsx-a11y/role-has-required-aria-props"
 - "jsx-a11y/role-supports-aria-props"
 - "jsx-a11y/scope"
 - "jsx-a11y/tabindex-no-positive"
</details>
<br>

## References
 - https://code.visualstudio.com/api/advanced-topics/tslint-eslint-migration
 - https://github.com/typescript-eslint/tslint-to-eslint-config/blob/master/docs/FAQs.md
 - https://github.com/microsoft/eslint-plugin-sdl
 - https://github.com/dustinspecker/awesome-eslint#plugins

## tslint-to-eslint-config.log
Here's the full summary from tslint-to-eslint-config:

<details>
    <summary>Expand</summary>
✨ 40 rules replaced with their ESLint equivalents. ✨

❗ 4 ESLint rules behave differently from their TSLint counterparts ❗
  Check ./tslint-to-eslint-config.log for details.

❓ 42 rules are not known by tslint-to-eslint-config to have ESLint equivalents. ❓
  The "@typescript-eslint/tslint/config" section of ./.eslintrc.js configures eslint-plugin-tslint to run them in TSLint within ESLint.
  Check ./tslint-to-eslint-config.log for details.

☠ Prettier plugins are missing from your configuration. ☠
  We highly recommend running tslint-to-eslint-config --prettier to disable formatting ESLint rules.
  See https://github.com/typescript-eslint/tslint-to-eslint-config/blob/master/docs/FAQs.md#should-i-use-prettier.

⚡ 5 new packages are required for this ESLint configuration. ⚡
  npm install @typescript-eslint/eslint-plugin @typescript-eslint/eslint-plugin-tslint @typescript-eslint/parser eslint eslint-plugin-import --save-dev

♻ Consider using --comments to replace TSLint comment directives in your source files. ♻

✅ All is well! ✅
    
Full log:
4 ESLint rules behave differently from their TSLint counterparts:

  * brace-style:
    - ESLint's brace-style will check all tokens.

  * no-underscore-dangle:
    - Leading or trailing underscores (_) on identifiers will now be forbidden.

  * @typescript-eslint/no-unused-expressions:
    - The TSLint optional config "allow-new" is the default ESLint behavior and will no longer be ignored.

  * eqeqeq:
    - Option "smart" allows for comparing two literal values, evaluating the value of typeof and null comparisons.

42 rules are not known by tslint-to-eslint-config to have ESLint equivalents:
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "jquery-deferred-must-complete".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "match-default-export-name".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "mocha-avoid-only".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "mocha-no-side-effect-code".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-delete-expression".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-disable-auto-sanitization".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-document-domain".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-document-write".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-exec-script".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-function-constructor-with-string-args".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-http-string".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-inner-html".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-jquery-raw-elements".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-string-based-set-immediate".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-string-based-set-interval".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-string-based-set-timeout".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-unnecessary-bind".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-unnecessary-override".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-unsupported-browser-code".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-with-statement".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "non-literal-require".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "possible-timing-attack".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-a11y-anchors".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-a11y-aria-unsupported-elements".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-a11y-event-has-role".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-a11y-image-button-has-alt".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-a11y-img-has-alt".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-a11y-lang".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-a11y-meta".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-a11y-props".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-a11y-proptypes".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-a11y-role".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-a11y-role-has-required-aria-props".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-a11y-role-supports-aria-props".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-a11y-tabindex-no-positive".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-a11y-titles".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-anchor-blank-noopener".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-iframe-missing-sandbox".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-no-dangerous-html".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-this-binding-issue".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "react-unused-props-and-state".
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "whitespace".
</details>
<br>

### CODE REVIEW

File: webapp/src/marked.tsx [hunk 1]
```diff
@@ -342,9 +342,7 @@ export class MarkedContent extends data.Component<MarkedContentProps, MarkedCont
         markdown = markdown.replace(/<\s*script[^>]*>.*<\/\s*script\s*>/g, '');
 
         // Render the markdown and add it to the content div
-        /* tslint:disable:no-inner-html (marked content is already sanitized) */
         content.innerHTML = marked(markdown);
```
Comment (riknoll) [inline webapp/src/marked.tsx:345]: This rule?
 // eslint-disable-next-line @microsoft/sdl/no-inner-html

File: webapp/src/blocks.tsx [hunk 2]
```diff
@@ -1744,9 +1744,7 @@ export class Editor extends toolboxeditor.ToolboxEditor {
                     let type = shadow.getAttribute('type');
                     const builtin = snippets.allBuiltinBlocks()[type];
                     let b = this.getBlockXml(builtin ? builtin : { name: type, attributes: { blockId: type } }, ignoregap, true);
-                    /* tslint:disable:no-inner-html setting one element's contents to the other */
```
Comment (riknoll) [inline webapp/src/blocks.tsx:1747]: This one should also use that SDL rule

**Comment by riknoll** [comment]
So is there a rule checking inner-html? imo we should keep that one around

File: pxtlib/docsrender.ts [hunk 3]
```diff
@@ -3,7 +3,8 @@
 /// <reference path="commonutil.ts"/>
 
 namespace pxt.docs {
-    declare var require: any;
+    // eslint-disable-next-line no-var
+declare var require: any;
```
Comment (shakao) [inline pxtlib/docsrender.ts:7]:
```suggestion
    declare var require: any;
```

File: webapp/src/container.tsx [hunk 4]
```diff
@@ -613,7 +612,7 @@ export class MainMenu extends data.Component<ISettingsProps, {}> {
                 </a> : undefined}
             </div>
         </div>;
-        /* tslint:enable:react-a11y-anchors */
+        /* eslint-enable jsx-a11y/anchor-is-valid */
```
Comment (shakao) [inline webapp/src/container.tsx:615]: should there be a matching "disable" for this "enable" ?

**Comment by shakao** [comment]
eslint caused a bunch of weirdness when i was first setting up the skillmap build, does everything still work if you run `pxt skillmap` in pxt-arcade?

**Comment by abchatra** [comment]
This is awesome @darzu. Kudos!

File: webapp/src/onedrive.ts [hunk 5]
```diff
@@ -417,6 +417,7 @@ export class Provider extends cloudsync.ProviderBase implements cloudsync.Provid
                 ifr.style.border = "0";
                 hiddenFrame = (document.getElementsByTagName("body")[0].appendChild(ifr) as HTMLIFrameElement);
             } else if (document.body && document.body.insertAdjacentHTML) {
+                // eslint-disable-next-line @microsoft/sdl/no-inner-html
```
Comment (abchatra) [inline webapp/src/onedrive.ts:420]: This seems to be a new disable. Is this safe in this context?

**Review by abchatra — APPROVED** [review]

File: webapp/src/onedrive.ts (hunk 5 above)
Comment (abchatra) [reply webapp/src/onedrive.ts:420]: Inner html are risky

File: webapp/src/blocks.tsx (hunk 2 above)
Comment (darzu) [reply webapp/src/blocks.tsx:1747]: > So is there a rule checking inner-html? imo we should keep that one around

Yes, there is a rule and these should all be brought back. I only realized there was a replacement rule (it wasn't found by the auto-migration tool) party way through the port so I missed some of these that I had already removed. I'll scan for all of these.

**Comment by darzu** [comment]
> eslint caused a bunch of weirdness when i was first setting up the skillmap build, does everything still work if you run `pxt skillmap` in pxt-arcade?

Good question, I'll try it out!

File: webapp/src/onedrive.ts (hunk 5 above)
Comment (darzu) [reply webapp/src/onedrive.ts:420]: I'll review it. We are using innerHtml in ~12 places :/

Comment (abchatra) [reply webapp/src/onedrive.ts:420]: @pelikhan who have found issues here in the past.

**Comment by pelikhan** [comment]
That entire file can go away? Time to clean up the old cloud sync stuff.

________________________________

**Comment by darzu** [comment]
> That entire file can go away? Time to clean up the old cloud sync stuff.

Yup, as soon as we're finished with micro:bit release I'm doing a cleanup of the workspace layer which includes getting rid of this old OneDrive stuff.

**Comment by abchatra** [comment]
I am not sure onedrive is linked to workspace cleanup. It can go out now. It is not used anywhere.

**Comment by darzu** [comment]
# Additional Changes

## Update node from v8 to v14
eslint wasn't compiling with node v8 because it depended on newer features. So I took a stab at updating node to v14 (in our package.json and github workflow builds). This seems to have worked everywhere except for our package "webfonts-generator".

## webfonts-generator
"webfonts-generator" started failing to build with node 14. This npm package was abandoned (https://www.npmjs.com/package/webfonts-generator) at our version, but there is a maintained fork (https://www.npmjs.com/package/@vusion/webfonts-generator) here that I switched us to and now everything builds, lints, and tests fine.

## Skillsmap, react-scripts & ESLint
Skillsmap uses react-scripts which internally uses eslint, but it was using eslint v6.x (this PR uses v7.x). For whatever reason I couldn't quite understand, react-scripts is unhappy about this even if I told it to disable linting. So I updated react-scripts to v4.x which uses v7.x and it no longer complains about incompatible versions. However, due to somewhat strange behavior in how eslint config files work (file paths are considered relative to the execution cwd of eslint proc, not the .eslintrc.js file), I had to disable eslint as part of the react-script build. However skillmap/ folder is still linted correctly by vscode and by `gulp lint`, just not by `pxt skillmap`. FYI: @riknoll , @shakao

File: webapp/src/container.tsx (hunk 4 above)
Comment (darzu) [reply webapp/src/container.tsx:615]: Hmm, it's not complaining so maybe we don't need the enable either. The rules aren't exactly 1-1

File: webapp/src/blocks.tsx (hunk 2 above)
Comment (darzu) [reply webapp/src/blocks.tsx:1747]: Ah so this is one of the two files that crashes eslint today. After this PR I'm going to upgrade our TS version which I think will fix the crash, then it'll start linting and requiring directives here. For now, they wouldn't do anything.

File: webapp/src/marked.tsx (hunk 1 above)
Comment (darzu) [reply webapp/src/marked.tsx:345]: For some reason that rule doesn't complain here. Maybe this is actually safe? Or the rule got less good at checking :/

Comment (darzu) [reply webapp/src/marked.tsx:345]: Ooooh, it looks like innerHTML is actually safe on `<div>`, `<span>`, and `<noembed>`.
TIL: https://developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML

File: docs/static/streamer/streamer.ts [hunk 6]
```diff
@@ -366,9 +366,7 @@ function onYouTubeIframeAPIReady() {
 
     function loadToolbox() {
         const config = readConfig();
-        // tslint:disable-next-line: no-inner-html
         toolbox.innerHTML = "";
-        // tslint:disable-next-line: no-inner-html
         paintbox.innerHTML = "";
```
Comment (darzu) [inline docs/static/streamer/streamer.ts:370]: Note that the new no-inner-html rule allows assigning the empty string since that is always safe.

**Comment by darzu** [comment]
# Additional Changes pt.2

## lint streamer
Streamer wasn't being linted by gulp but probably was being linted by Peli's tslint vscode plugin. The eslint vscode plugin doesn't lint files unless they're part of a .eslintrc.js tree, so I added a .eslintrc.js file to streamer and added it to the gulp lint build. It's worth noting that a lot of the rules, esp the new no-inner-html rule, are more precise and permissive than before. All the innerHtml usage in streamer was actually safe and no longer needs a lint exception.

## removed onedrive.ts
Wasn't used.

**Review by jwunderl — APPROVED** [review]
You love to see it 
image

**Pull request merged**